ast/file_info.go

```diff
@@ -155,13 +161,84 @@ func (f *FileInfo) isDummyFile() bool {
 	return f.lines == nil
 }
 
+// FirstToken returns the first token in the file. It returns
+// TokenError if the file contains no tokens (e.g. it is empty
+// or contains only whitespace and comments).
+func (f *FileInfo) FirstToken() Token {
+	return f.tokenForward(0)
+}
+
+// NextToken returns the next token in the file that comes after
+// t. It returns TokenError if there is no next token (i.e. the
+// given t is the last token). It also returns TokenError if
+// the given t is invalid.
+func (f *FileInfo) NextToken(t Token) Token {
+	if t < 0 {
+		return TokenError
+	}
```

Comment:
I originally had a `t++` in `tokenForward`, which is why I separated this out: `FirstToken` was always passing -1.

Though I can't really _move_ the check -- it would have to be replicated (since we'd also want to check the original `t`, not just check if `t` overflowed.

Maybe I a better check is to add an upper bound check here (as well as adding a lower bound check in `PreviousToken`), so that we avoid overflow/underflow that way.

Reply:
Added extra range checks.